File: python/src/graph.py
"""Road network graph construction and adjacency index for route optimization.

Constructs a weighted directed graph from raw node and edge datasets, executing
full physics derivation on each edge at load time. Provides a unified interface
for cost weight retrieval across all three optimization modes.
"""

from typing import List, Dict, Tuple
from models import Node, Edge, Vehicle, Config
from physics import compute_edge_metrics


class Graph:
    """Weighted directed road network graph with pre-computed physical edge costs.

    At instantiation, each edge undergoes a complete vehicle physics evaluation
    (aerodynamic drag, slope gradient, rolling resistance, fuel energy conversion),
    yielding all three objective cost fields (FASTEST, CHEAPEST, BALANCED).
    The resulting graph is immutable after construction and safe for concurrent reads.

    Attributes:
        nodes: Dictionary mapping node ID to Node geometry.
        edges: Ordered list of all directed road segments.
        adj: Adjacency mapping from node ID to a list of (target_id, Edge) tuples.
        vehicle: Vehicular physical constants used during edge metric computation.
        config: Atmospheric and optimization weight parameters.
    """

    __slots__ = ('nodes', 'adj', 'edges', 'vehicle', 'config')

    def __init__(self, nodes: List[Node], edges: List[Edge], vehicle: Vehicle, config: Config) -> None:
        """Initializes the graph, triggers physics derivation, and builds adjacency structures.

        Args:
            nodes: Ordered list of spatial graph vertices.
            edges: Ordered list of directed road segment entities.
            vehicle: Vehicular aerodynamic and thermodynamic constants.
            config: Atmospheric parameters and multi-objective weight coefficients.
        """
        self.nodes: Dict[int, Node] = {n.id: n for n in nodes}
        self.edges: List[Edge] = edges
        self.vehicle: Vehicle = vehicle
        self.config: Config = config

        # Build forward adjacency list: node_id -> [(target_id, edge), ...]
        self.adj: Dict[int, List[Tuple[int, Edge]]] = {n.id: [] for n in nodes}
        for e in edges:
            compute_edge_metrics(e, vehicle, config)
            if e.source in self.adj:
                self.adj[e.source].append((e.target, e))

    def get_edge_weight(self, edge: Edge, mode: str) -> float:
        """Returns the pre-computed objective cost for the specified optimization mode.

        Args:
            edge: The road segment whose cost is to be retrieved.
            mode: Optimization strategy identifier. One of:
                - ``"FASTEST"``  – minimizes travel time (minutes).
                - ``"CHEAPEST"`` – minimizes total monetary expenditure (fuel + tolls).
                - ``"BALANCED"`` – evaluates weighted combination of all three objectives.

        Returns:
            Scalar edge traversal cost for the priority queue relaxation step.
        """
        if mode == "FASTEST":
            return edge.cost_fastest
        elif mode == "CHEAPEST":
            return edge.cost_cheapest
        elif mode == "BALANCED":
            return edge.cost_balanced
        else:
            return edge.cost_balanced

File: rust/src/graph.rs
//! Weighted directed road network graph with pre-computed physical edge costs.
//!
//! Constructs an indexed adjacency structure over a set of spatial vertices and
//! directed road segments. All edge physical metrics (aerodynamic drag, slope force,
//! rolling resistance, fuel consumption, and per-mode costs) are evaluated once at
//! construction time via [`crate::physics::compute_edge_metrics`], making the graph
//! read-only and safe for concurrent multi-threaded query execution via `Rayon`.

use crate::models::{Config, Edge, Node, Vehicle};
use crate::physics::compute_edge_metrics;
use std::collections::HashMap;

/// Weighted directed road network graph with pre-computed multi-objective edge costs.
///
/// Adjacency entries store `(target_node_id, edge_index)` pairs rather than direct
/// edge references, keeping the adjacency list compact and enabling cache-efficient
/// edge lookup during priority queue relaxation steps.
#[derive(Debug, Clone)]
pub struct Graph {
    /// Mapping from node identifier to spatial Node geometry.
    pub nodes: HashMap<usize, Node>,
    /// Ordered vector of all directed road segments with synthesized cost fields.
    pub edges: Vec<Edge>,
    /// Vehicular physical constants used during edge metric computation.
    pub vehicle: Vehicle,
    /// Atmospheric constants and multi-objective weight parameters.
    pub config: Config,
    /// Forward adjacency index: `node_id → [(target_node_id, edge_index_in_edges)]`.
    pub adj: HashMap<usize, Vec<(usize, usize)>>,
}

impl Graph {
    /// Constructs the graph, executes full physics derivation on all edges,
    /// and builds the forward adjacency index.
    ///
    /// # Arguments
    /// * `nodes_list` - Complete list of spatial graph vertices.
    /// * `edges_list` - Complete list of directed road segments (mutated in-place).
    /// * `vehicle`    - Vehicular aerodynamic and thermodynamic constants.
    /// * `config`     - Atmospheric parameters and optimization weight coefficients.
    ///
    /// # Returns
    /// A fully constructed `Graph` ready for concurrent route query execution.
    pub fn new(
        nodes_list: Vec<Node>,
        mut edges_list: Vec<Edge>,
        vehicle: Vehicle,
        config: Config,
    ) -> Self {
        let mut nodes = HashMap::with_capacity(nodes_list.len());
        let mut adj: HashMap<usize, Vec<(usize, usize)>> = HashMap::with_capacity(nodes_list.len());

        for n in nodes_list {
            adj.insert(n.id, Vec::new());
            nodes.insert(n.id, n);
        }

        for (i, edge) in edges_list.iter_mut().enumerate() {
            compute_edge_metrics(edge, &vehicle, &config);
            if let Some(neighbors) = adj.get_mut(&edge.source) {
                neighbors.push((edge.target, i));
            }
        }

        Self {
            nodes,
            edges: edges_list,
            vehicle,
            config,
            adj,
        }
    }

    /// Returns the pre-computed objective cost for the specified optimization mode.
    ///
    /// # Arguments
    /// * `edge` - The road segment whose cost field is to be selected.
    /// * `mode` - Optimization strategy: `"FASTEST"`, `"CHEAPEST"`, or `"BALANCED"`.
    ///
    /// # Returns
    /// Scalar edge traversal cost for use in priority queue relaxation.
    pub fn get_edge_weight(&self, edge: &Edge, mode: &str) -> f64 {
        match mode {
            "FASTEST"  => edge.cost_fastest,
            "CHEAPEST" => edge.cost_cheapest,
            _          => edge.cost_balanced,
        }
    }
}
